burr/core/application.py

```diff
@@ -1127,38 +1181,152 @@ def with_hooks(self, *adapters: LifecycleAdapter) -> "ApplicationBuilder":
         `storage_dir` and `app_id`, which have automatic defaults.
 
-        :param project: Project name
-        :param tracker: Tracker to use, currently only ``local`` is available
-        :param params: Parameters to pass to the tracker
+        :param tracker: Tracker to use. ``local`` creates one, else pass one in.
+        :param project: Project name -- used if the tracker is local.
+        :param params: Parameters to pass to the tracker if it's local.
         :return: The application builder for future chaining.
         """
         if params is None:
             params = {}
-        if tracker == "local":
+        # if it's a lifecycle adapter, just add it
+        if isinstance(
+            tracker,
+            (
+                PostRunStepHook,
+                PreRunStepHook,
+                PostApplicationCreateHook,
+                PostEndSpanHook,
+                PreStartSpanHook,
+            ),
+        ):
+            self.lifecycle_adapters.append(tracker)
+        elif tracker == "local":
             from burr.tracking.client import LocalTrackingClient
 
-            self.lifecycle_adapters.append(LocalTrackingClient(project=project, **params))
+            kwargs = {"project": project, "app_id": self.app_id}
+            kwargs.update(params)
+            self.lifecycle_adapters.append(LocalTrackingClient(**kwargs))
+        else:
+            raise ValueError(f"Tracker {tracker}:{project} not supported")
+        return self
+
+    def initialize_from(
+        self,
+        persister: BasicStatePersistence,
+        resume_at_next_action: bool,
+        default_state: dict,
+        default_entry_point: str,
+    ) -> "ApplicationBuilder":
+        self.persister = persister
+        self.resume_at_next_action = resume_at_next_action
+        self.default_state = default_state
+        self.start = default_entry_point
+        return self
+
+    def with_state_persister(
+        self, persister: BasicStatePersistence, on_every: str = "step"
+    ) -> "ApplicationBuilder":
+        """Adds a state persister to the application. This is a way to persist state out to a database, file, etc...
+        at the specified interval.
+
+        :param persister: The persister to add
+        :param on_every: The interval to persist state. Currently only "step" is supported.
+        :return: The application builder for future chaining.
+        """
+        if on_every != "step":
+            raise ValueError(f"on_every {on_every} not supported")
+        # stick persister into a post_step_hook if it doesn't implement it already.
+        if hasattr(persister, "post_step_hook"):
+            if persister not in self.lifecycle_adapters:
+                self.lifecycle_adapters.append(persister)
         else:
-            raise ValueError(f"Tracker {tracker} not supported")
+            # TODO: pull this out more formally
+            class persisterWrapper(PostRunStepHook):
```

Comment:
Class capitalization?

Reply:
Define this elsewhere -- this is muddying up the application , which is already way too long. Probably should live with the persister.